### ISSUE TITLE
nix-shell does not come with cabal ?

### ISSUE THREAD

**Issue by PierreR** [dhall-lang/dhall-haskell#2024]
I am trying to give some feedback about contribution and what could be improved to make it easier.

This one comes a bit as a surprise to me:

If you go in the dhall-haskell/dhall folder and execute `nix-shell` then if you try `cabal build` it does not work:

```
[nix-shell:~/projects/hub/dhall-haskell/dhall]$ cabal build
cabal: command not found
```

Isn't this supposed to work ? nix-shell should give you an development environment with everything required to contribute.

I would like to reach a point where contributing is easy. You just do `code` to open vscode and you have an IDE that just works (with the help of ghcide or [haskell-language-server](https://github.com/haskell/[haskell-language-server)).

See this post: https://daml.com/daml-driven/handing-over-ghcide-to-the-haskell-community/

Sorry if I miss the point. I am just a dumb user at this point ;-)

**Comment by PierreR** [comment]
Something else that would be quite useful is to have a shell that not only includes cabal but also comes with a setup that makes cabal uses the system dependencies from nix but **also** all the project dependencies.

IIRC this can be achieved with `pkgs.haskellPackages.shellFor`

**Comment by Gabriel439** [comment]
@PierreR: Regarding the latter request, that is how the `nix-shell` already works.  It provides all dependencies, including the Haskell ones.

**Comment by PierreR** [comment]
@Gabriel439  are you sure cabal uses them when doing a `cabal build` ?

**Comment by Gabriel439** [comment]
@PierreR: As an extra precaution, try running `cabal configure` again when you enter the `nix-shell`.  That will help ensure that `cabal` uses the `ghc` that Nix provides, which has all of the Nix-supplied packages in the GHC package database

**Comment by PierreR** [comment]
@Gabriel439 I am missing something ... If I don't provide a ghc I would get the following:
```
 ~/projects/hub/dhall-haskell  master ✓ 
→ nix-shell
Dhall core tools shell
USAGE EXAMPLES
    Default (GHC version 8.6.5, option completions shell - bash):
        $ nix-shell
    Overriding default GHC compiler (see ./nix/shared.nix for available options):
        $ nix-shell --arg args '{ compiler = ghc861; }'
    Overriding option completion shell flavor:
        $ nix-shell --argstr shell zsh
dhall-docs: command not found

[nix-shell:~/projects/hub/dhall-haskell]$ ghc --version
ghc: command not found
```

**Comment by Gabriel439** [comment]
@PierreR: Oh, the `nix-shell` is supposed to be run from within the project subdirectory.  See the [Nix](https://github.com/dhall-lang/dhall-haskell/blob/master/README.md#nix) section of the `README` for more details

**Comment by PierreR** [comment]
@Gabriel439 Thanks ! Now I have a ghc ;-) I still need to figure out why `cabal` is not included in the shell. 

I have also found out that nix-build does not build the same haskell package deps than nix-shell ... (maybe because I am not using the same cabal ?). 

Another surprised is that after building dhall within a shell on the dhall folder, if  I go to the dhall-openapi directory and enter nix-shell then it is going to rebuild dhall again.

**Comment by PierreR** [comment]
Passing the initial surprise about `nix-shell` being incomplete (`cabal` is missing) I have managed to find my way by installing a version of `ghcide` that matches the ghc version (using nix to install cabal in user space).

I am a bit surprised that each project folder seems to require a full nix rebuild of Dhall when entering the shell. I would expect the dhall derivation to be the same across `dhall-docs` and `dhall-openapi` for instance

It would be great to be able to start from the root, enter nix-shell, type 'ghcide' and be able to "jump to definition" from one dhall project to another.

A new issue could be created to enhance the first time experience if it is something of interest.

Thanks for your help @Gabriel439 !

**Comment by Gabriel439** [comment]
I'm reopening because there is still something that I can do here related to the original request of providing `cabal` and/or `vscode` with a `ghcide` plugin

**Comment by PierreR** [comment]
Trying to fiddle with `dhall-openapi` today and that was my experience:
 
t -1:  the setup was working nicely a week ago

1. go to the repo/ git pull/ rebase the change on your local branches ...
2. go to dhall-openapi, open vscode, well that's really optimistic
3. I need to `nix-shell` -> wait 5 min because dhall 1.35.0 has to be rebuilt
(it recompiles twice -- why I wonder )
4. ghcide -> boom -> problem with hnix (strange that is not even a dependency of `dhall-openapi`)
5. take a few minutes to realize I need to do `cabal update`
6 cabal build just to be sure - wait it rebuilds `dhall` again (ok let's wait)
7. success !
8. ghcide 
```
> Building library for dhall-1.35.0..
> gcc: error: /home/vagrant/projects/hub/dhall-haskell/dist-newstyle/build/x86_64-linux/ghc-8.6.5/dhall-1.35.0/build/Dhall/Parser/Token.dyn_o: No such file or directory
> `cc' failed in phase `Linker'. (Exit code: 1)
```
Wait what ?
9. cabal clean
10. ghcide -> let's wait for ... `dhall-1.35.0` to rebuild
11. success !! 😌
😇

**Comment by Gabriel439** [comment]
@PierreR: Yeah, Nixpkgs builds Haskell packages with profiling enabled by default (although that's configurable)

**Comment by timbertson** [comment]
I am very confused, because suddenly I'm also having the original issue - `cabal` not present in a nix shell (I'm inside the `dhall` subdirectory). It was present previously, I don't even know what I did. I have checked out a new version and done `nix-channel --update`, but I tried both checking out the previous version and also rolling back my nixpkgs. I believe nixpkgs is pinned anyway, so that _shouldn't_ affect it 😕

**Comment by Gabriel439** [comment]
@timbertson: The `nix-shell` currently does not provide `cabal` (this is an idiosyncracy of how Nix's Haskell support works).  If `cabal` was ever present it would have to have been supplied out-of-band

**Comment by timbertson** [comment]
Ah, I didn't realise that. Now that I think of it, I suspect I edited a shell.nix early on to provide this, and it got lost in a `git checkout` somewhere.

Now I've got this rather hacky version in a standalone file (`dhall/cabal-shell.nix`), so it's less likely to be clobbered:

```
$ cat cabal-shell.nix
let base = (import ../nix/shared.nix {}); in
base.dhall.env.overrideAttrs (o: {
  buildInputs = o.buildInputs ++ [ base.pkgs.cabal-install ];
})
```

I am curious what the expected development flow is, though. Do you have `cabal` installed globally? Would you be open to a flag to provide `cabal` within the standard shell? Is the main downside that it's unnecessary for those who already have a global install?

**Comment by Gabriel439** [comment]
Yes, I install `cabal` globally for my own dev workflow, but I'd accept a pull request to have the shell include `cabal`.

We don't even need to have a flag: I'd accept adding it to the shell unconditionally because there's already an easy way to opt out, which is to use the `nix: True` flag.  This lets you use `cabal` outside of a `nix-shell` so that you can still use your globally installed version.

The main downside that I'm aware of is that `cabal` doesn't like multiple different versions of `cabal` using the same package index file (which is shared across all uses of `cabal` for any given user).  However, it automatically recovers from this by regenerating the index with a warning (and it takes some time).

**Comment by sjakobi** [comment]
This was apparently fixed in #2283.
